# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd

## Regression Model Evaluation

In [2]:
from sklearn.datasets import load_boston

data = load_boston()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=['MEDV'])

data = pd.concat([X, y], axis=1)

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [3]:
from sklearn.model_selection import train_test_split

In [4]:
X_train, X_test, y_train, y_test = train_test_split( X, 
                                                    y, 
                                                    test_size=0.20, 
                                                    random_state=123)

## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [5]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [6]:
lr = LinearRegression()

lr.fit(X_train, y_train)

y_train_pred = lr.predict(X_train)

y_test_pred = lr.predict(X_test)

## 3. Calculate and print R-squared for both the training and the testing set.

In [7]:
r2 = [round(r2_score(y_train, y_train_pred),3),round(r2_score(y_test, y_test_pred),3) ]
r2

[0.756, 0.659]

## 4. Calculate and print mean squared error for both the training and the testing set.

In [8]:
from sklearn.metrics import mean_squared_error as mse

In [9]:
mean_square = [round(mse(y_train, y_train_pred),3), round(mse(y_test, y_test_pred),3)]
mean_square

[20.677, 28.192]

## 5. Calculate and print mean absolute error for both the training and the testing set.

In [10]:
from sklearn.metrics import mean_absolute_error as mae

In [11]:
abs_error = [round(mae(y_train, y_train_pred),3), round(mae(y_test, y_test_pred),3)]

abs_error

[3.175, 3.66]

#### The r2 score for the training set is greater than the score for the testing set, and the MSE is lower also lower in the former. If these were independent models we could say that the first one is better. 
#### In this case they belong to the same set and model, with just differentiating between trainin and test, it's reasonable to see that there are predicitions done on the training set are more accurate that those of the testing set, as the model has been trained with those same values. Also, the mean absolute error is lower in the training set. 
#### Having said this, the model is not at all bad, with an r2 of 65.9%, a MSE of 28,19 and an MAE of 3.66. 

## Classification Model Evaluation

In [36]:
from sklearn.datasets import load_iris

data = load_iris()

X = pd.DataFrame(data["data"], columns=data["feature_names"])
y = pd.DataFrame(data["target"], columns=["class"])

data = pd.concat([X, y], axis=1)

In [37]:
y["class"].value_counts()
# There are three classes, so we should use a multiclass aproach 

2    50
1    50
0    50
Name: class, dtype: int64

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [13]:
X_train, X_test, y_train, y_test = train_test_split( X, 
                                                    y, 
                                                    test_size=0.20)

## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [14]:
from sklearn.linear_model import LogisticRegression

In [16]:
lg = LogisticRegression(max_iter = 1000)

lg.fit(X_train, y_train.values.ravel()) # A warning appeared about the size of the y_train and suggested using .raven()

y_train_pred = lg.predict(X_train)

y_test_pred = lg.predict(X_test)


## 8. Calculate and print the accuracy score for both the training and the testing set.

In [17]:
from sklearn.metrics import accuracy_score

In [21]:
training_accuracy = accuracy_score(
    y_true=y_train,
    y_pred=y_train_pred
)

testing_accuracy = accuracy_score(
    y_true=y_test,
    y_pred=y_test_pred
)

results = [round(training_accuracy, 3), round(testing_accuracy,3)]

results

[0.967, 0.967]

## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [20]:
from sklearn.metrics import balanced_accuracy_score as bas

In [23]:
balanced_accuracy = [round(bas(y_train, y_train_pred),3), round(bas(y_test, y_test_pred),3)]
balanced_accuracy

[0.964, 0.978]

## 10. Calculate and print the precision score for both the training and the testing set.

In [25]:
from sklearn.metrics import precision_score as ps

In [41]:
precision_score = [round(ps(y_train, y_train_pred, average = 'weighted'),3), 
                   round(ps(y_test, y_test_pred, average = 'weighted'),3)]
precision_score

[0.967, 0.972]

## 11. Calculate and print the recall score for both the training and the testing set.

In [45]:
from sklearn.metrics import recall_score as rs

In [50]:
recall_score = [round(rs(y_train, y_train_pred, average = 'weighted'), 3), 
                round(rs(y_test, y_test_pred, average = 'weighted'),3)]
recall_score

[0.967, 0.967]

## 12. Calculate and print the F1 score for both the training and the testing set.

In [51]:
from sklearn.metrics import f1_score as f1

In [53]:
f_one = [round(f1(y_train, y_train_pred, average = "weighted"),3),
         round(f1(y_test, y_test_pred, average = 'weighted'),3)]
f_one

[0.967, 0.968]

## 13. Generate confusion matrices for both the training and the testing set.

In [54]:
from sklearn.metrics import confusion_matrix

In [55]:
confusion_matrix(y_train, y_train_pred)

array([[40,  0,  0],
       [ 0, 44,  1],
       [ 0,  3, 32]], dtype=int64)

In [56]:
confusion_matrix(y_test, y_test_pred)

array([[10,  0,  0],
       [ 0,  5,  0],
       [ 0,  1, 14]], dtype=int64)

## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.